helm/resource_release.go

```diff
@@ -603,54 +610,94 @@ func resourceReleaseCreate(ctx context.Context, d *schema.ResourceData, meta int
-	client.Devel = d.Get("devel").(bool)
-	client.DependencyUpdate = d.Get("dependency_update").(bool)
-	client.Timeout = time.Duration(d.Get("timeout").(int)) * time.Second
-	client.Namespace = d.Get("namespace").(string)
-	client.ReleaseName = d.Get("name").(string)
-	client.GenerateName = false
-	client.NameTemplate = ""
-	client.OutputDir = ""
-	client.Atomic = d.Get("atomic").(bool)
-	client.SkipCRDs = d.Get("skip_crds").(bool)
-	client.SubNotes = d.Get("render_subchart_notes").(bool)
-	client.DisableOpenAPIValidation = d.Get("disable_openapi_validation").(bool)
-	client.Replace = d.Get("replace").(bool)
-	client.Description = d.Get("description").(string)
-	client.CreateNamespace = d.Get("create_namespace").(bool)
+	var rel *release.Release
+	var releaseAlreadyExists bool
+	var enableUpgradeStrategy bool
 
-	if cmd := d.Get("postrender.0.binary_path").(string); cmd != "" {
-		av := d.Get("postrender.0.args")
-		var args []string
-		for _, arg := range av.([]interface{}) {
-			if arg == nil {
-				continue
-			}
-			args = append(args, arg.(string))
-		}
+	releaseName := d.Get("name").(string)
+	enableUpgradeStrategy = d.Get("upgrade_install").(bool)
 
-		pr, err := postrender.NewExec(cmd, args...)
-		if err != nil {
+	if enableUpgradeStrategy {
+		// Check to see if there is already a release installed.
+		histClient := action.NewHistory(actionConfig)
+		histClient.Max = 1
+		if _, err := histClient.Run(releaseName); errors.Is(err, driver.ErrReleaseNotFound) {
+			debug("%s Chart %s is not yet installed", logID, chartName)
+		} else if err != nil {
 			return diag.FromErr(err)
+		} else {
+			releaseAlreadyExists = true
+			debug("%s Chart %s is installed as release %s", logID, chartName, releaseName)
 		}
-
-		client.PostRenderer = pr
 	}
 
-	debug("%s Installing chart", logID)
+	if enableUpgradeStrategy && releaseAlreadyExists {
+		debug("%s Upgrade-installing chart installed out of band", logID)
 
-	rel, err := client.Run(c, values)
+		upgradeClient := action.NewUpgrade(actionConfig)
+		upgradeClient.ChartPathOptions = *cpo
+		upgradeClient.DryRun = false
+		upgradeClient.DisableHooks = d.Get("disable_webhooks").(bool)
+		upgradeClient.Wait = d.Get("wait").(bool)
+		upgradeClient.Devel = d.Get("devel").(bool)
+		upgradeClient.Timeout = time.Duration(d.Get("timeout").(int)) * time.Second
+		upgradeClient.Namespace = d.Get("namespace").(string)
+		upgradeClient.Atomic = d.Get("atomic").(bool)
+		upgradeClient.SkipCRDs = d.Get("skip_crds").(bool)
+		upgradeClient.SubNotes = d.Get("render_subchart_notes").(bool)
+		upgradeClient.DisableOpenAPIValidation = d.Get("disable_openapi_validation").(bool)
+		upgradeClient.Description = d.Get("description").(string)
 
+		if cmd := d.Get("postrender.0.binary_path").(string); cmd != "" {
+			pr, err := postrender.NewExec(cmd)
+			if err != nil {
+				return diag.FromErr(err)
+			}
+			upgradeClient.PostRenderer = pr
+		}
+
+		debug("%s Upgrading chart", logID)
```

Comment:
From running it manually once more I noticed the following flow:

1. Install `bitnami/redis` chart with `version = "19.4.0"`
2. Remove tfstate + remove `version = "19.4.0"`
3. `terraform init` -> `terraform apply`
4. It proceeds to perform an upgrade to the latest version. `19.6.1`

This is something that we'll want to address since this goes against what was discussed prior:
> This is good news since this PR should **only** support `helm upgrade --install`, auto upgarding to the latest release is not something we want to support in the provider. This is noted [here](https://github.com/hashicorp/terraform-provider-helm/issues/102#issuecomment-621311452)

I did test every other case and it works as expected, such as:

>1. installed `bitnami/redis` with no version set, marking the version as the latest (`19.6.1`)
>2. explicitly add `version = 19.6.0`, changing the version `19.6.1` -> `19.6.0`
>3. remove `version` to see whether the version is upgrade to `19.6.1`This results in no changes outputted by terraform

I'm requesting changes here since it seems like this would be where we'd want to check if the release exists and if it does use the version associated with the already existing release in order to prevent an accidental upgrade to the latest release. Let me know if you have any questions, we appreciate your effort in this! @n-oden

Reply:
I don't love the provider having explicitly different semantics than the default helm tooling, but c'est la guerre; I'll try to get this sorted today.

Reply:
Actually... I'm going to push back on this a little, because what you are asking for is directly in opposition to what your own documentation suggests is the expected behavior, which is that if the `version` field is absent, the latest version will be installed:

```go
			"version": {
				Type:        schema.TypeString,
				Optional:    true,
				Computed:    true,
				Description: "Specify the exact chart version to install. If this is not specified, the latest version is installed.",
			},
```

...which is in alignment with how the helm cli works: if the `--version` flag is not passed in and the chart is being installed from a repo, the latest version will be picked, and this is true for both `helm install` and `helm upgrade`.

Changing this behavior in the case where `upgrade_install=true` seems to me like it breaks the principle of least surprise, and I think it would be better to explicitly warn in the documentation that the combination of upgrade_install=true and version=null will potentially change the installed release.

*That all said* commit 17fc4fd7a does what you've requested and adds test coverage for it.

Reply:
The `version` description was a mistake on my part, this was actually brought up in the past where the documentation itself was updated. Unfortunately I did not update the description in the schema itself 😓 https://github.com/hashicorp/terraform-provider-helm/issues/924

In the past, users would try to create a `helm_release` resource where the release would _already_ be present which where they would get this panic:

```hcl
│ Error: cannot re-use a name that is still in use
│ 
│   with helm_release.upgrade,
│   on main.tf line 1, in resource "helm_release" "upgrade":
│    1: resource "helm_release" "upgrade" {
│ 
```
the ideal way to resolve this would be to run `terraform import` in order to get the already installed release into the tfstate, this also prevents any unexpected changes to the release itself with this error.

The `upgrade-install` would also resolve this error but we still want to promote being explicit in regards to version bumping, something that's been mentioned in the past https://github.com/hashicorp/terraform-provider-helm/issues/102#issuecomment-621311452

> This is the expected behaviour - we want to encourage the practice of being explicit about the upgrade version when upgrading an installed chart. We'll take a note to improve the documentation referenced above.